sw/device/tests/aes_sideload_test.c

```diff
@@ -0,0 +1,290 @@
+  AES_TESTUTILS_WAIT_FOR_STATUS(&aes, kDifAesStatusInputReady, true, TIMEOUT);
+  CHECK_DIF_OK(dif_aes_load_data(&aes, in_data_plain));
+
+  // Trigger the AES decryption and wait for it to complete.
+  CHECK_DIF_OK(dif_aes_trigger(&aes, kDifAesTriggerStart));
+  AES_TESTUTILS_WAIT_FOR_STATUS(&aes, kDifAesStatusOutputValid, true, TIMEOUT);
+
+  // Verify that the output is equal to the plain text.
+  dif_aes_data_t out_data_plain;
+  CHECK_DIF_OK(dif_aes_read_output(&aes, &out_data_plain));
+  CHECK_ARRAYS_EQ(out_data_plain.data, kPlainText, ARRAYSIZE(kPlainText));
+
+  // Finish the ECB decryption transaction.
+  CHECK_DIF_OK(dif_aes_end(&aes));
+
+  LOG_INFO("Clearing the side loaded key.");
+
+  // Clear the key in the keymgr and decrypt the ciphertext again.
+  CHECK_DIF_OK(
+      dif_keymgr_sideload_clear_set_enabled(&keymgr, kDifToggleEnabled));
+  CHECK_DIF_OK(dif_aes_start(&aes, &transaction, NULL, NULL));
+
+  AES_TESTUTILS_WAIT_FOR_STATUS(&aes, kDifAesStatusInputReady, true, TIMEOUT);
+  CHECK_DIF_OK(dif_aes_load_data(&aes, in_data_plain));
+
+  // Trigger the AES decryption and wait for it to complete.
+  CHECK_DIF_OK(dif_aes_trigger(&aes, kDifAesTriggerStart));
+  AES_TESTUTILS_WAIT_FOR_STATUS(&aes, kDifAesStatusOutputValid, true, TIMEOUT);
+
+  // Verify that output is not equal to the plain text.
+  CHECK_DIF_OK(dif_aes_read_output(&aes, &out_data_plain));
+  CHECK_ARRAYS_NE(out_data_plain.data, kPlainText, ARRAYSIZE(kPlainText));
+  CHECK_DIF_OK(dif_aes_end(&aes));
+}
+
+static void init_kmac_for_keymgr(void) {
+  dif_kmac_t kmac;
+  CHECK_DIF_OK(
+      dif_kmac_init(mmio_region_from_addr(TOP_EARLGREY_KMAC_BASE_ADDR), &kmac));
+
+  // Configure KMAC hardware using software entropy.
+  CHECK_DIF_OK(dif_kmac_configure(&kmac, (dif_kmac_config_t){
+                                             .sideload = true,
+                                         }));
+}
+
+bool test_main(void) {
+  dif_rstmgr_t rstmgr;
+  dif_rstmgr_reset_info_bitfield_t info;
+
+  CHECK_DIF_OK(dif_rstmgr_init(
+      mmio_region_from_addr(TOP_EARLGREY_RSTMGR_AON_BASE_ADDR), &rstmgr));
+  CHECK_DIF_OK(dif_rstmgr_reset_info_get(&rstmgr, &info));
+
+  init_flash();
+
+  // POR reset.
+  if (info == kDifRstmgrResetInfoPor) {
+    LOG_INFO("Powered up for the first time, program flash");
+
+    // Lock otp secret partition.
+    dif_otp_ctrl_t otp;
+    CHECK_DIF_OK(dif_otp_ctrl_init(
+        mmio_region_from_addr(TOP_EARLGREY_OTP_CTRL_CORE_BASE_ADDR), &otp));
+    otp_ctrl_testutils_lock_partition(&otp, kDifOtpCtrlPartitionSecret2, 0);
+
+    // Reboot device.
+    CHECK_DIF_OK(dif_rstmgr_reset_info_clear(&rstmgr));
+    CHECK_DIF_OK(dif_rstmgr_software_device_reset(&rstmgr));
+
+    // Wait here until device reset.
+    wait_for_interrupt();
+
+  } else if (info == kDifRstmgrResetInfoSw) {
+    LOG_INFO(
+        "Powered up for the second time, actuate keymgr and perform aes test.");
+
+    init_kmac_for_keymgr();
```

Comment:
If I remove that kmac initialization only the dvsim will fail during the keymgr advance to `kDifKeymgrStateCreatorRootKey` , FPGA and Veriltor will pass. Is that expected?
@sriyerg @vogelpi Do you know?

Reply:
I don't know why it only results in a failure for Xcelium/VCS. Maybe it triggers an assertion? These are not active in Verilator. Also, the FPGA uses the unmasked KMAC version due to resource constraints.

Reply:
@weicaiyang to help debug the simulation with VCS - could you provide more details regarding the error seen?

Reply:
do you want me to have a look? @weicaiyang ?

Reply:
Eunchan and I just looked into this. It's due to an assertion error. I filed an issue - https://github.com/lowRISC/opentitan/issues/13594